src/app/src/main/kotlin/org/vaccineimpact/api/app/serialization/DataTableDeserializer.kt

```diff
@@ -0,0 +1,157 @@
+package org.vaccineimpact.api.app.serialization
+
+import com.opencsv.CSVReader
+import org.vaccineimpact.api.Deserializer
+import org.vaccineimpact.api.app.errors.ValidationError
+import org.vaccineimpact.api.models.ErrorInfo
+import org.vaccineimpact.api.models.helpers.FlexibleColumns
+import java.io.StringReader
+import kotlin.reflect.KClass
+import kotlin.reflect.KFunction
+import kotlin.reflect.KType
+import kotlin.reflect.full.findAnnotation
+import kotlin.reflect.full.primaryConstructor
+
+class HeaderDefinition(val name: String, val type: KType)
+
+open class DataTableDeserializer<out T>(
+        protected val headers: List<HeaderDefinition>,
+        private val constructor: KFunction<T>
+)
+{
+    private val headerCount = headers.size
+    protected open val extraHeadersAllowed = false
+
+    fun deserialize(body: String): Sequence<T>
+    {
+        val reader = CSVReader(StringReader(body.trim()))
+
+        var row = reader.readNext()
+        val headers = row.toList()
+        checkHeaders(headers)
+        val actualHeaders = getActualHeaderDefinitions(headers)
```

Comment:
Totally agree about `headers` being a bad name for that `val`. I've renamed it to `headerDefinitions`.

I'm not sure about the second point. `actualHeaders` *does* mean the headers returned in the CSV. It just so happens that in the simple case of a completely fixed data structure this will be the same as the expected header definitions, as we've already checked they match up in `checkHeaders`. I've renamed `val headers = row.toList()` to `val actualHeaderNames...`. Is that any better?